keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+---
+kep-number: 13
+title: pull based audit events stream
+authors:
+  - "@CaoShuFeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-11
+last-updated: 2018-06-11
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+
+---
+
+# Pull based audit events stream
+
+## Table of Contents
+
+* [Pull based audit events stream](#pull-based-audit-events-stream)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Make audit backends dynamicly registerable](#make-audit-backends-dynamicly-registerable)
+    * [endpoints and query parameters](#endpoints-and-query-parameters)
+    * [Bandwidth limit Configuration](#bandwidth-limit-configuration)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
+
+## Motivation
+The current audit backends (log & webhook) require very high privileges (file access to the master) or prior configuration (webhook). There are a number of use cases for dynamic and unprivileged access to the audit logs. Examples include:
+- allow tenants to get audit events in their own namespace
+- provide a subset of audit events, so it will be easier to query and analyze (e.g. https://github.com/kubernetes/kubernetes/issues/56683)
+
+
+### Goals
+- Provide a new pull based audit event stream, which returns events that occurred after the stream was opened.
+- Allow users go get a subset of audit events.
+- Access control. Allow unprivileged users to get audits in their own namespace.
```

Comment:
How do we make sure that the audit events do not leak information the user cannot access via "normal" resource queries otherwise (i.e. which are restricted via RBAC) ?

Reply:
We can have such level restriction:
Allow A to read audit events in namespace-1.

We can't have this:
Allow A to read audit events `about pods,servers` in namespace-1.

Reply:
Since we treat audit events as `pseudo-resource`. I think we can restrict the access control of audit events as other normal resources.

But we don't support `resourceNames`, since audit events don't have a name.

Reply:
They could have a pseudo-name.

I was wondering whether we really want to invent another authz system for audit events or whether we can piggy pack on RBAC rules. I.e. when querying (via /audit or a real resource URL) I see all events of objects I could also read via normal APIs. Or we add another verb to RBAC called "AUDIT". Just thinking out loud.

/cc @kubernetes/sig-auth-misc @liggitt

Reply:
I think for an initial implementation it's acceptable to just authorize it at the audit resource level (i.e. I need read permissions on the `audit/namespaces/foo` resource), and document this as a something to watch out for. Then it's up to the admin / namespace admin to restrict audit permissions to those who should actually be able to read everything in the cluster/namespace.

Also, sensitive information (secrets) should be filtered from audit logs anyway.

Reply:
I mean this should just be handled in the policy. e.g. https://github.com/kubernetes/kubernetes/blob/56d9d73f38489f09d55ccee442ed77169484b146/cluster/gce/gci/configure-helper.sh#L837-L839. The whole object body is excluded in this case.